csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/BotServices.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+
+using System;
+using System.Collections.Generic;
+using Microsoft.ApplicationInsights;
+using Microsoft.Bot.Builder.AI.Luis;
+using Microsoft.Bot.Builder.AI.QnA;
+
+namespace NLP_With_Dispatch_Bot
+{
+    /// <summary>
+    /// Represents the bot's references to external services.
+    ///
+    /// For example, Application Insights, Luis models and QnaMaker services
+    /// are kept here (singletons).  These external services are configured
+    /// using the BotConfigure class (based on the contents of your ".bot" file).
+    /// </summary>
+    [Serializable]
+    public class BotServices
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BotServices"/> class.
+        /// </summary>
+        /// <param name="client">An Application Insights <see cref="TelemetryClient"/> instance.</param>
+        /// <param name="qnaServices">A dictionary of named <see cref="QnAMaker"/> instances for usage within the bot.</param>
+        public BotServices(TelemetryClient client, Dictionary<string, QnAMaker> qnaServices, Dictionary<string, LuisRecognizer> luisServices)
+        {
+            TelemetryClient = client ?? throw new ArgumentNullException(nameof(client));
+            QnAServices = qnaServices ?? throw new ArgumentNullException(nameof(qnaServices));
+            LuisServices = luisServices ?? throw new ArgumentNullException(nameof(luisServices));
+        }
+
+        /// <summary>
+        /// Gets the Application Insights Telemetry client.
+        /// Use this to log new custom events/metrics/traces/etc into your
+        /// Application Insights service for later analysis.
+        /// </summary>
+        /// <value>
+        /// The Application Insights <see cref="TelemetryClient"/> instance created based on configuration in the .bot file.
+        /// </value>
+        public TelemetryClient TelemetryClient { get; }
+
+        /// <summary>
+        /// Gets the (potential) set of QnA Services used.
+        /// Given there can be multiple QnA services used in a single bot,
+        /// QnA is represented as a Dictionary.  This is also modeled in the
+        /// ".bot" file since the elements are named (string).
+        /// This sample only uses a single QnA instance.
+        /// </summary>
+        /// <value>
+        /// A QnAMaker client instance created based on configuration in the .bot file.
+        /// </value>
+        public Dictionary<string, QnAMaker> QnAServices { get; } = new Dictionary<string, QnAMaker>();
```

Comment:
>Dictionary [](start = 15, length = 10)

I wonder if this need to be a concurrent dictionary? We set this up at construction time, at which point its immutable. Multiple threads then read from it. 

... however if one of those threads tired to change this (aka: Dynamic QnA Services) then there's a problem. 

I *think* it's ok as-is, but I could be easily persuaded that this should be a concurrent dictionary.

Reply:
I think this is OK, we don't want to teach to over-complicate things. The static QnaMaker (or set of QnaMakers) will be the case for 99.999 of bots. Good catch though, since it is not really immutable.

Reply:
We could add a comment perhaps?

// We use Dictionary for our services, since this dictionary is set up at class construction and then remains immutable. 
// In a scenario where QnaServices or LuisServices change at runtime, make sure to switch to a thread-safe collection.